File: android/src/main/java/ru/innim/flutter_login_vk/MethodCallHandler.java
package ru.innim.flutter_login_vk;

import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import com.vk.api.sdk.BuildConfig;
import com.vk.api.sdk.VK;
import com.vk.api.sdk.VKApiCallback;
import com.vk.api.sdk.auth.VKAccessToken;
import com.vk.api.sdk.auth.VKScope;
import com.vk.sdk.api.account.AccountService;
import com.vk.sdk.api.users.UsersService;
import com.vk.sdk.api.users.dto.UsersFields;
import com.vk.sdk.api.users.dto.UsersUserXtrCounters;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.Result;

public class MethodCallHandler implements MethodChannel.MethodCallHandler {
    private final static String _LOGIN_METHOD = "logIn";
    private final static String _LOGOUT_METHOD = "logOut";
    private final static String _GET_ACCESS_TOKEN = "getAccessToken";
    private final static String _GET_USER_PROFILE = "getUserProfile";
    private final static String _GET_SDK_VERSION = "getSdkVersion";
    private final static String _INIT_SDK_METHOD = "initSdk";
    private final static String _SCOPE_LOGIN_ARG = "scope";
    private final static String _SCOPE_INIT_ARG = "scope";

    private final LoginCallback _loginCallback;
    private Activity _activity;
    private Context _context;

    public MethodCallHandler(Context context, LoginCallback loginCallback) {
        _loginCallback = loginCallback;
        _context = context;
    }

    public void updateActivity(Activity activity) {
        _activity = activity;
    }

    @Override
    public void onMethodCall(MethodCall call, Result r) {
        if (_activity != null) {
            switch (call.method) {
                case _LOGIN_METHOD:
                    final List<String> scope = call.argument(_SCOPE_LOGIN_ARG);
                    logIn(scope, r);
                    break;
                case _LOGOUT_METHOD:
                    logOut();
                    result(null, r);
                    break;
                case _GET_ACCESS_TOKEN:
                    result(getAccessToken(), r);
                    break;
                case _GET_USER_PROFILE:
                    getUserProfile(r);
                    break;
                case _GET_SDK_VERSION:
                    result(getSdkVersion(), r);
                    break;
                case _INIT_SDK_METHOD:
                    final List<String> initScope = call.argument(_SCOPE_INIT_ARG);
                    initSdk(initScope, r);
                    break;
                default:
                    r.notImplemented();
                    break;
            }
        }
    }

    private void initSdk(final List<String> scope, final Result r) {
        VK.initialize(_context);

        if (scope != null && VK.isLoggedIn()) {
            final int userId = VK.getUserId();
            VK.execute(new AccountService().accountGetAppPermissions(userId), new VKApiCallback<Integer>() {
                @Override
                public void success(Integer o) {
                    List<String> list = Arrays.asList(scope.toArray(new String[0]));
                    List<VKScope> vkScopes = getScopes(list);
                    if (!VKClient.hasScopes(vkScopes, o)) {
                        VK.logout();
                    }

                    result(true, r);
                }

                @Override
                public void fail(@NotNull Exception o) {
                    error(FlutterError.apiError("Get profile permissions error: " + o.getMessage(),
                            new VKError(0, o.getMessage())), r);
                }
            });
        } else {
            result(true, r);
        }
    }

    private void logIn(List<String> scopes, Result result) {
        _loginCallback.addPending(result);

        List<String> list = Arrays.asList(scopes.toArray(new String[0]));
        List<VKScope> vkScopes = getScopes(list);

        VK.login(_activity, vkScopes);
    }

    private List<VKScope> getScopes(List<String> list) {
        List<VKScope> vkScopes = new ArrayList<>();
        int count = list.size();
        for (int i = 0; i < count; ++i) {
            String item = list.get(i);
            VKScope scope = VKScope.valueOf(item.toUpperCase());
            vkScopes.add(scope);
        }

        return vkScopes;
    }

    private void logOut() {
        VK.logout();
    }

    private HashMap<String, Object> getAccessToken() {
        if (VK.isLoggedIn()) {
            final VKAccessToken token = VKClient.getCurrentAccessToken();
            if (token != null) {
                return Results.accessToken(token);
            }
        }

        return null;
    }

    private void getUserProfile(final Result r) {
        List<UsersFields> fields = VKClient.FIELDS_DEFAULT;

        final VKAccessToken token = VKClient.getCurrentAccessToken();
        if (token != null) {
            VK.execute(new UsersService().usersGet(null, fields, null),
                    new VKApiCallback<List<UsersUserXtrCounters>>() {
                        @Override
                        public void success(List<UsersUserXtrCounters> o) {
                            result(Results.userProfile(o.get(0)), r);
                        }

                        @Override
                        public void fail(@NotNull Exception o) {
                            error(FlutterError.apiError("Get profile error: " + o.getMessage(),
                                    new VKError(0, o.getMessage())), r);
                        }
                    });
        }
    }

    private String getSdkVersion() {
        return BuildConfig.VERSION_NAME;
    }

    private void result(Object data, Result r) {
        r.success(data);
    }

    private void error(FlutterError error, Result r) {
        r.error(error.code, error.message, error.details);
    }
}

File: android/src/main/kotlin/ru/innim/flutter_login_vk/MethodCallHandler.kt
package ru.innim.flutter_login_vk

import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import com.vk.api.sdk.*
import com.vk.api.sdk.VK.execute
import com.vk.api.sdk.VK.getUserId
import com.vk.api.sdk.VK.initialize
import com.vk.api.sdk.VK.isLoggedIn
import com.vk.api.sdk.VK.login
import com.vk.api.sdk.VK.logout
import com.vk.api.sdk.auth.VKAccessToken
import com.vk.api.sdk.auth.VKScope
import com.vk.sdk.api.account.AccountService
import com.vk.sdk.api.users.UsersService
import com.vk.sdk.api.users.dto.UsersUserFull
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.*

class MethodCallHandler(private val context: Context, private val loginCallback: LoginCallback)
    : MethodChannel.MethodCallHandler {
    companion object {
        private const val logIn = "logIn"
        private const val logOut = "logOut"
        private const val getAccessToken = "getAccessToken"
        private const val getUserProfile = "getUserProfile"
        private const val getSdkVersion = "getSdkVersion"
        private const val initSdk = "initSdk"
        private const val scopeLoginArg = "scope"
        private const val scopeInitArg = "scope"
    }

    private var activity: Activity? = null

    fun updateActivity(activity: Activity?) {
        this.activity = activity
    }

    override fun onMethodCall(call: MethodCall, r: MethodChannel.Result) {
        activity?.let {
            when (call.method) {
                logIn -> {
                    val scope = call.argument<List<String>>(scopeLoginArg) ?: listOf()
                    logIn(scope, r)
                }
                logOut -> {
                    logOut()
                    sendResult(null, r)
                }
                getAccessToken -> sendResult(getAccessToken(), r)
                getUserProfile -> getUserProfile(r)
                getSdkVersion -> sendResult(getSdkVersion(), r)
                initSdk -> {
                    val initScope = call.argument<List<String>?>(scopeInitArg)
                    initSdk(initScope, r)
                }
                else -> r.notImplemented()
            }
        }
    }

    private fun initSdk(scope: List<String>?, channelResult: MethodChannel.Result) {
        initialize(context)

        if (scope != null && isLoggedIn()) {
            val userId = getUserId()
            execute(AccountService().accountGetAppPermissions(userId), object : VKApiCallback<Int> {
                override fun success(result: Int) {
                    val list = listOf(*scope.toTypedArray())
                    val vkScopes: List<VKScope> = getScopes(list)
                    if (!VKClient.hasScopes(vkScopes, result)) {
                        logout()
                    }
                    sendResult(true, channelResult)
                }

                override fun fail(error: Exception) {
                    sendError(FlutterError.apiError("Get profile permissions error: " + error.message,
                            VKError(0, error.message.toString())), channelResult)
                }
            })
        } else {
            sendResult(true, channelResult)
        }
    }

    private fun logIn(scopes: List<String>, result: MethodChannel.Result) {
        loginCallback.addPending(result)
        val list = listOf(*scopes.toTypedArray())
        val vkScopes: List<VKScope> = getScopes(list)
        // TODO: use ActivityResultLauncher
        // There are multiple problems with this new login approach:
        // 1. It's requires ComponentActivity, so FlutterActivity is not applicable
        // 2. We need register login activity at start of the app.
        // Which is not crucial but looks pretty bad.
        // 3. Even using FlutterFragmentActivity, we get error
        // "It's requires ComponentActivity, so FlutterActivity is not applicable" on launch
        @Suppress("DEPRECATION")
        login(activity!!, vkScopes)
    }

    private fun getScopes(list: List<String>): List<VKScope> {
        val vkScopes: MutableList<VKScope> = ArrayList()
        val count = list.size
        for (i in 0 until count) {
            val item = list[i]
            val scope = VKScope.valueOf(item.uppercase())
            vkScopes.add(scope)
        }
        return vkScopes
    }

    private fun logOut() {
        logout()
    }

    private fun getAccessToken(): HashMap<String, Any?>? {
        if (isLoggedIn()) {
            val storage = VKPreferencesKeyValueStorage(context)
            val token = VKAccessToken.restore(storage)

            if (token != null) {
                return Results.accessToken(token)
            }
        }

        return null
    }

    private fun getUserProfile(r: MethodChannel.Result) {
        val fields = VKClient.fieldsDefault

        execute(UsersService().usersGet(fields = fields),
                object : VKApiCallback<List<UsersUserFull?>?> {
                    override fun success(result: List<UsersUserFull?>?) {
                        if (result != null && result.isNotEmpty() && result[0] != null) {
                            sendResult(Results.userProfile(result[0]!!), r)
                        } else {
                            sendError(FlutterError.invalidResult("Get profile error: the result is null"), r)
                        }
                    }

                    override fun fail(error: Exception) {
                        sendError(FlutterError.apiError("Get profile error: " + error.message,
                                VKError(0, error.message.toString())), r)
                    }
                })
    }

    private fun getSdkVersion(): String {
        val metaData = context.packageManager.getApplicationInfo(
                context.packageName, PackageManager.GET_META_DATA)
        return metaData.metaData["VKSdkVersion"].toString()
    }

    private fun sendResult(data: Any?, r: MethodChannel.Result) {
        r.success(data)
    }

    private fun sendError(error: FlutterError, r: MethodChannel.Result) {
        r.error(error.code, error.message, error.details)
    }
}
